[jit] improve interface error messages

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25228 [jit] improve interface error messages**
* #25227 [jit] add serialization of interface
* #25226 Remove PythonPrint's is_method_ member
* #25258 Add interface declarations to JIT

This adds a facility to isSubtypeOf for it to explain why a type is
not a subtype of something else. It is used in situations where it
is not clear from the types python_str alone why the relationship
is now true. Because of subtle interaction between default arguments,
overloads, and virtual methods, it uses isSubtypeOfExt for the extended
version to avoid requiring readers to understand the interaction.

Differential Revision: [D17066673](https://our.internmc.facebook.com/intern/diff/D17066673)

zdevito committed:
[jit] improve interface error messages

This adds a facility to isSubtypeOf for it to explain why a type is
not a subtype of something else. It is used in situations where it
is not clear from the types python_str alone why the relationship
is now true. Because of subtle interaction between default arguments,
overloads, and virtual methods, it uses isSubtypeOfExt for the extended
version to avoid requiring readers to understand the interaction.

zdevito committed:
Update on "[jit] improve interface error messages"

[jit] improve interface error messages

This adds a facility to isSubtypeOf for it to explain why a type is
not a subtype of something else. It is used in situations where it
is not clear from the types python_str alone why the relationship
is now true. Because of subtle interaction between default arguments,
overloads, and virtual methods, it uses isSubtypeOfExt for the extended
version to avoid requiring readers to understand the interaction.

gh-metadata: pytorch pytorch 25228 gh/zdevito/96/head

zdevito committed:
Update on "[jit] improve interface error messages"

[jit] improve interface error messages

This adds a facility to isSubtypeOf for it to explain why a type is
not a subtype of something else. It is used in situations where it
is not clear from the types python_str alone why the relationship
is now true. Because of subtle interaction between default arguments,
overloads, and virtual methods, it uses isSubtypeOfExt for the extended
version to avoid requiring readers to understand the interaction.

gh-metadata: pytorch pytorch 25228 gh/zdevito/96/head

zdevito committed:
Update on "[jit] improve interface error messages"

[jit] improve interface error messages

This adds a facility to isSubtypeOf for it to explain why a type is
not a subtype of something else. It is used in situations where it
is not clear from the types python_str alone why the relationship
is now true. Because of subtle interaction between default arguments,
overloads, and virtual methods, it uses isSubtypeOfExt for the extended
version to avoid requiring readers to understand the interaction.

gh-metadata: pytorch pytorch 25228 gh/zdevito/96/head

zdevito committed:
Update on "[jit] improve interface error messages"

[jit] improve interface error messages

This adds a facility to isSubtypeOf for it to explain why a type is
not a subtype of something else. It is used in situations where it
is not clear from the types python_str alone why the relationship
is now true. Because of subtle interaction between default arguments,
overloads, and virtual methods, it uses isSubtypeOfExt for the extended
version to avoid requiring readers to understand the interaction.

gh-metadata: pytorch pytorch 25228 gh/zdevito/96/head

ZolotukhinM reviewed (approved):
Looks good modulo some formatting issues.

zdevito committed:
Update on "[jit] improve interface error messages"

[jit] improve interface error messages

This adds a facility to isSubtypeOf for it to explain why a type is
not a subtype of something else. It is used in situations where it
is not clear from the types python_str alone why the relationship
is now true. Because of subtle interaction between default arguments,
overloads, and virtual methods, it uses isSubtypeOfExt for the extended
version to avoid requiring readers to understand the interaction.

gh-metadata: pytorch pytorch 25228 gh/zdevito/96/head
